File: apps/mobile/modules/centraid-ocr/android/src/main/java/expo/modules/centraidocr/CentraidOcrModule.kt
package expo.modules.centraidocr

import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

private const val MAX_BYTES = 25L * 1024L * 1024L
private const val MAX_PIXELS = 20_000_000L

class CentraidOcrModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CentraidOcr")

    AsyncFunction("recognizeText") { rawUri: String, promise: Promise ->
      try {
        val uri = Uri.parse(rawUri)
        require(uri.scheme == "file") { "OCR accepts local file URLs only." }
        val file = File(requireNotNull(uri.path))
        require(file.length() <= MAX_BYTES) {
          "Scan exceeds the 25 MB on-device OCR limit."
        }
        val context = requireNotNull(appContext.reactContext)
        val image = InputImage.fromFilePath(context, uri)
        require(image.width.toLong() * image.height.toLong() <= MAX_PIXELS) {
          "Scan exceeds the 20 megapixel on-device OCR limit."
        }
        val recognizer =
          TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        recognizer.process(image)
          .addOnSuccessListener { result ->
            val lines = result.textBlocks.flatMap { block ->
              block.lines.map { line ->
                mapOf(
                  "text" to line.text,
                  // The bundled Latin recognizer does not expose a stable
                  // per-line confidence, so report extraction confidence as
                  // present/absent and keep every field reviewable.
                  "confidence" to if (line.text.isBlank()) 0.0 else 1.0
                )
              }
            }
            val confidence =
              if (lines.isEmpty()) 0.0
              else lines.map { it["confidence"] as Double }.average()
            promise.resolve(
              mapOf(
                "text" to result.text,
                "confidence" to confidence,
                "lines" to lines,
                "engine" to "ml-kit"
              )
            )
            recognizer.close()
          }
          .addOnFailureListener { error ->
            recognizer.close()
            promise.reject("ERR_OCR", error.message, error)
          }
      } catch (error: Throwable) {
        promise.reject("ERR_OCR", error.message, error)
      }
    }
  }
}

File: apps/mobile/modules/centraid-ocr/ios/CentraidOcrModule.swift
import ExpoModulesCore
import Foundation
import UIKit
import Vision

private let maxBytes: UInt64 = 25 * 1024 * 1024
private let maxPixels = 20_000_000

public class CentraidOcrModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CentraidOcr")

    AsyncFunction("recognizeText") { (uri: String) throws -> [String: Any] in
      let url = try localFileUrl(uri)
      let values = try url.resourceValues(forKeys: [.fileSizeKey])
      guard UInt64(values.fileSize ?? 0) <= maxBytes else {
        throw OcrError("Scan exceeds the 25 MB on-device OCR limit.")
      }
      guard
        let image = UIImage(contentsOfFile: url.path),
        let cgImage = image.cgImage
      else {
        throw OcrError("The selected file is not a readable image.")
      }
      guard cgImage.width * cgImage.height <= maxPixels else {
        throw OcrError("Scan exceeds the 20 megapixel on-device OCR limit.")
      }

      let request = VNRecognizeTextRequest()
      request.recognitionLevel = .accurate
      request.usesLanguageCorrection = true
      try VNImageRequestHandler(cgImage: cgImage).perform([request])
      let lines: [[String: Any]] = (request.results ?? []).compactMap { item in
        guard let candidate = item.topCandidates(1).first else { return nil }
        return [
          "text": candidate.string,
          "confidence": Double(candidate.confidence)
        ]
      }
      let confidence = lines.isEmpty
        ? 0
        : lines.reduce(0) { $0 + (($1["confidence"] as? Double) ?? 0) }
          / Double(lines.count)
      return [
        "text": lines.compactMap { $0["text"] as? String }.joined(separator: "\n"),
        "confidence": confidence,
        "lines": lines,
        "engine": "apple-vision"
      ]
    }
  }
}

private func localFileUrl(_ raw: String) throws -> URL {
  guard let url = URL(string: raw), url.isFileURL else {
    throw OcrError("OCR accepts local file URLs only.")
  }
  return url
}

private struct OcrError: Error, LocalizedError {
  let message: String
  init(_ message: String) { self.message = message }
  var errorDescription: String? { message }
}
